swarm_addr.go

```diff
@@ -9,16 +11,58 @@ import (
 func (s *Swarm) ListenAddresses() []ma.Multiaddr {
 	s.listeners.RLock()
 	defer s.listeners.RUnlock()
+	return s.listenAddressesNoLock()
+}
+
+func (s *Swarm) listenAddressesNoLock() []ma.Multiaddr {
 	addrs := make([]ma.Multiaddr, 0, len(s.listeners.m))
 	for l := range s.listeners.m {
 		addrs = append(addrs, l.Multiaddr())
 	}
 	return addrs
 }
 
+const ifaceAddrsCacheDuration = 1 * time.Minute
+
 // InterfaceListenAddresses returns a list of addresses at which this swarm
 // listens. It expands "any interface" addresses (/ip4/0.0.0.0, /ip6/::) to
 // use the known local interfaces.
 func (s *Swarm) InterfaceListenAddresses() ([]ma.Multiaddr, error) {
-	return addrutil.ResolveUnspecifiedAddresses(s.ListenAddresses(), nil)
+	s.listeners.RLock() // RLock start
+
+	ifaceListenAddres := s.listeners.ifaceListenAddres
+	isEOL := time.Now().After(s.listeners.cacheEOL)
+	s.listeners.RUnlock() // RLock end
+
+	if ifaceListenAddres != nil && !isEOL {
+		// Cache is valid
+		return ifaceListenAddres, nil
```

Comment:
We may want to check if our listeners have changed. I'm a bit worried users will call `Listen` and then wonder why that address isn't listed here.

Reply:
Done.